zk_watcher.go

```diff
@@ -382,6 +382,42 @@ func (w *zkWatcher) createAll(node string) error {
 	return nil
 }
 
+// triggerCleanup walks the prefix and deletes any non-ephemeral, empty
+// znodes. It swallows any errors encountered.
+func (w *zkWatcher) triggerCleanup() {
+	w.RLock()
+	defer w.RUnlock()
+
+	w.cleanupTree(w.prefix)
+}
+
+func (w *zkWatcher) cleanupTree(node string) bool {
+	children, stat, err := w.conn.Children(node)
+	if err != nil {
+		return false
+	} else if stat.EphemeralOwner() != 0 {
+		return false
+	}
+
+	empty := true
+	for _, child := range children {
+		if !w.cleanupTree(path.Join(node, child)) {
+			empty = false
+		}
+	}
+
+	if empty {
```

Comment:
I think we'd want to check for that exact situation in this block anyways, so we may as well eliminate the other logic?